src/mame/machine/agatkeyb.h

```diff
@@ -0,0 +1,82 @@
+// license:BSD-3-Clause
+// copyright-holders:Sergey Svishchev
+#ifndef MAME_MACHINE_AGATKEYB_H
+#define MAME_MACHINE_AGATKEYB_H
+
+#pragma once
+
+#include "machine/keyboard.h"
+
+
+/***************************************************************************
+    DEVICE TYPE GLOBALS
+***************************************************************************/
+
+DECLARE_DEVICE_TYPE(AGAT_KEYBOARD, agat_keyboard_device)
+
+
+/***************************************************************************
+    REUSABLE I/O PORTS
+***************************************************************************/
+
+INPUT_PORTS_EXTERN( agat_keyboard );
+
+
+/***************************************************************************
+    TYPE DECLARATIONS
+***************************************************************************/
+
+
+class agat_keyboard_device : public device_t, protected device_matrix_keyboard_interface<6U>
+{
+public:
+	typedef device_delegate<void (u16 character)> output_delegate;
+
+	agat_keyboard_device(
+		const machine_config &mconfig,
+		const char *tag,
+		device_t *owner,
+		u32 clock);
+
+	template <typename... T>
+	void set_keyboard_callback(T &&... args)
+	{
+		m_keyboard_cb.set(std::forward<T>(args)...);
+	}
+
+	virtual ioport_constructor device_input_ports() const override;
+
+	auto out_meta_callback() { return m_out_meta_cb.bind(); }
+
+	DECLARE_INPUT_CHANGED_MEMBER(meta_changed);
+
+protected:
+	agat_keyboard_device(
+		const machine_config &mconfig,
+		device_type type,
+		char const *tag,
+		device_t *owner,
+		u32 clock);
+	virtual void device_start() override;
+	virtual void device_reset() override;
+	virtual void key_make(u8 row, u8 column) override;
+	virtual void key_repeat(u8 row, u8 column) override;
+	virtual void send_key(u16 code);
```

Comment:
Does this really need to be virtual?  Are there more keyboards in the family that will only differ in how they send the codes?

Reply:
Probably not.   There are at least two more keyboards (almost certainly with different matrix), but I'm not sure they're interesting enough to emulate.

Reply:
Better not to make unnecessary vtable entries then.  The translate function probably doesn’t need to be virtual either.